public/app/features/query/components/QueryEditorRenderer/QueryEditorRenderer.tsx

```diff
@@ -0,0 +1,73 @@
+import React from 'react';
+import { useAsync } from 'react-use';
+
+import { CoreApp, DataQuery, PanelData, TimeRange } from '@grafana/data';
+import { getDataSourceSrv } from '@grafana/runtime';
+import { hasReactQueryEditor, ReactEditorRenderer } from './ReactEditorRenderer';
+import { AngularEditorRenderer, hasAngularQueryEditor } from './AngularEditorRenderer';
+
+interface Props {
+  nameOrUid: string;
+  onChange: (query: DataQuery) => void;
+  onRunQuery: () => void;
+  query: DataQuery;
+  queries: DataQuery[];
+  timeRange?: TimeRange;
+  data?: PanelData;
+  app?: CoreApp;
+}
+
+export function QueryEditorRenderer(props: Props) {
+  const { app, nameOrUid, timeRange, query, queries, onChange, data, onRunQuery } = props;
+  const state = useAsync(() => getDataSourceSrv().get(nameOrUid), [nameOrUid]);
```

Comment:
in QueryEditorRow this is handled on a layer above, QueryEditorRow is passed a DataSourceInstanceSettings

Reply:
Well, yes and no. The actual datasource is loaded in the `loadDatasource` function in `QueryEditorRow`.

Reply:
Aha, I know understand what you mean. I will do the proper refactorings and move the loading if the datasource to the alerting context.